packages/core/src/lib/connection_manager.ts

```diff
@@ -0,0 +1,360 @@
+      this.dialAttemptsForPeer.set(peerId.toString(), dialAttempt + 1);
+
+      //remove peer from peerStore if dial fails after max attempts
+      if (dialAttempt > maxDialAttemptsForPeer) {
+        try {
+          log(
+            "Deleting undialable peer " + peerId.toString() + "from peerStore"
+          );
+          await this.libp2pComponents.peerStore.delete(peerId);
+          return;
+        } catch (error) {
+          throw (
+            "Error deleting undialable peer" +
+            peerId.toString() +
+            " from peerStore"
+          );
+        }
+      }
+
+      log(
+        `Failed to dial ${peerId.toString()} - reattempting (${dialAttempt})`
+      );
+
+      this.dialPeer(peerId);
+    }
+  }
+
+  private startPeerDiscoveryListener(): void {
+    const onDiscovery = async (evt: CustomEvent<PeerInfo>): Promise<void> => {
+      const { id: peerId } = evt.detail;
+
+      if (!(await this.isPeerDialable(peerId))) return;
+
+      this.dialPeer(peerId).catch((err) =>
+        log(`Error dialing peer ${peerId.toString()} : ${err}`)
+      );
+    };
+
+    this.libp2pComponents.peerStore.addEventListener("peer", onDiscovery);
+  }
+
+  private startPeerConnectionListener(
+    keepAliveOptions: KeepAliveOptions,
+    relay?: IRelay
+  ): void {
+    const onConnect = (evt: CustomEvent<Connection>): void => {
+      {
+        this.startKeepAlive(
+          evt.detail.remotePeer,
+          this.libp2pComponents.ping.bind(this),
+          keepAliveOptions,
+          relay
+        );
+      }
+    };
+
+    this.libp2pComponents.connectionManager.addEventListener(
+      "peer:connect",
+      onConnect
+    );
+  }
+
+  private startPeerDisconnectionListener(): void {
+    // TODO: ensure that these following issues are updated and confirmed
+    /**
+     * NOTE: Event is not being emitted on closing nor losing a connection.
+     * @see https://github.com/libp2p/js-libp2p/issues/939
+     * @see https://github.com/status-im/js-waku/issues/252
+     *
+     * >This event will be triggered anytime we are disconnected from another peer,
+     * >regardless of the circumstances of that disconnection.
+     * >If we happen to have multiple connections to a peer,
+     * >this event will **only** be triggered when the last connection is closed.
+     * @see https://github.com/libp2p/js-libp2p/blob/bad9e8c0ff58d60a78314077720c82ae331cc55b/doc/API.md?plain=1#L2100
+     */
+    this.libp2pComponents.connectionManager.addEventListener(
+      "peer:disconnect",
+      (evt) => {
+        this.stopKeepAlive(evt.detail.remotePeer);
+      }
+    );
+  }
+
+  /**
+   * Checks if the peer is dialable based on the following conditions:
+   * 1. If the peer is a bootstrap peer, it is only dialable if the number of current bootstrap connections is less than the max allowed.
+   * 2. If the peer is not a bootstrap peer
+   */
+  private async isPeerDialable(peerId: PeerId): Promise<boolean> {
```

Comment:
```suggestion
  private async shouldDialPeer(peerId: PeerId): Promise<boolean> {
```

As per your comment, this functions returns whether you should dial a peer, not whether the peer itself is _dialable_. A bootstrap peer is _dialable_ but you just don't want to dial it.